src/NHibernate/Proxy/FieldInterceptorProxyBuilder.cs

```diff
@@ -0,0 +1,298 @@
+using System;
+using System.Collections.Generic;
+using System.Linq;
+using System.Reflection;
+using System.Reflection.Emit;
+using NHibernate.Intercept;
+using NHibernate.Util;
+
+namespace NHibernate.Proxy
+{
+	internal static class FieldInterceptorProxyBuilder
+	{
+		private const MethodAttributes constructorAttributes =
+			MethodAttributes.Public | MethodAttributes.HideBySig | MethodAttributes.SpecialName | MethodAttributes.RTSpecialName;
+
+		private static readonly ConstructorInfo ObjectConstructor = typeof(object).GetConstructor(System.Type.EmptyTypes);
+
+		private static readonly System.Type FieldInterceptorAccessorType = typeof(IFieldInterceptorAccessor);
+		private static readonly PropertyInfo AccessorTypeFieldInterceptorProperty =
+			FieldInterceptorAccessorType.GetProperty(nameof(IFieldInterceptorAccessor.FieldInterceptor));
+		private static readonly System.Type FieldInterceptorType = typeof(IFieldInterceptor);
+		private static readonly MethodInfo FieldInterceptorInterceptMethod = FieldInterceptorType.GetMethod(nameof(IFieldInterceptor.Intercept));
+		private static readonly MethodInfo FieldInterceptorMarkDirtyMethod = FieldInterceptorType.GetMethod(nameof(IFieldInterceptor.MarkDirty));
+		private static readonly System.Type AbstractFieldInterceptorType = typeof(AbstractFieldInterceptor);
+		private static readonly FieldInfo AbstractFieldInterceptorInvokeImplementationField =
+			AbstractFieldInterceptorType.GetField(nameof(AbstractFieldInterceptor.InvokeImplementation));
+
+		public static TypeInfo CreateProxyType(System.Type baseType)
+		{
+			// Avoid having a suffix ending with "Proxy", for disambiguation with INHibernateProxy proxies
+			var typeName = $"{baseType.Name}ProxyForFieldInterceptor";
```

Comment:
Previous implementation was causing field interceptor proxies to be build with same type and assembly name than `INHibernateProxy`. Apparently this was not causing any issue, even when lazily loading an entity as a `INHibernateProxy` loading its target implementation as a field interceptor proxy due to the entity having lazy properties.
Still it seems to me cleaner to avoid the confusion.